### ISSUE TITLE
Sentinel Mode : Error getting master address: ERR only (P)SUBSCRIBE / (P)UNSUBSCRIBE / PING / QUIT allowed in this context

### ISSUE THREAD

**Issue by adoignies** [vert-x3/vertx-redis-client#307]
### Version

Version 4.1.2

### Context

When requesting redis via the sentinel mode, the request works 10 times and then failed in error.

After investigation, the problem is caused by the master addr name retrieval:
`
conn.send(cmd(SENTINEL).arg("GET-MASTER-ADDR-BY-NAME").arg(masterName), getMasterAddrByName -> {
...});
`

The failed cause is "ERR only (P)SUBSCRIBE / (P)UNSUBSCRIBE / PING / QUIT allowed in this context".

My assumption is that the connection has also be used to subscribe to sentinel event, and redis reject the get-master-addr request.
the number 10 times is related to the defaut max pool size and the number of sentinels, when i set the max pool size to 3 then it failed after 4 success.

### Do you have a reproducer?

`
        RedisOptions redisOptions = new RedisOptions().setType(RedisClientType.SENTINEL);

        redisOptions.addConnectionString("redis:127.0.0.1:26378");
        redisOptions.addConnectionString("redis:127.0.0.1:26379");
        
        Redis redis = Redis.createClient(
                vertx,
                redisOptions
                        .setMasterName("masterName")
                        .setRole(RedisRole.MASTER));

        router.route().path("/redis").handler(c -> {
            c.request().pause();
            redis.send(Request.cmd(Command.INFO)).onComplete(info -> {
                c.request().resume();
                c.response().end("test resdis: " + info.failed());
            });

        });
`

### Steps to reproduce

1. Play ten times a redis command with a redis sentinel client...

### Extra

* with openjdk8

**Comment by adoignies** [comment]
If I reproduce it with a simple netcat connection

Connection to 127.0.0.1 26379 port [tcp/*] succeeded!
PING
+PONG
SUBSCRIBE +switch-master
*3
$9
subscribe
$14
+switch-master
:1

PING
*2
$4
pong
$0


sentinel GET-MASTER-ADDR-BY-NAME mastername
-ERR only (P)SUBSCRIBE / (P)UNSUBSCRIBE / PING / QUIT allowed in this context
`

**Comment by adoignies** [comment]
With further investigations, it seems that the subscribe request and the get-master-addr are done each time.
Therefore it seems the sentinelConnection used for subscribing is hang over back to the pool, and then used at the end for the get-master-addr and then resulting in the "only (P)SUBSCRIBE / (P)UNSUBSCRIBE / PING / QUIT allowed in this context".

For test, if I remove the subscribe part in RedisSentinelClient.connect, the behaviour dissapears. But the get-master-addr is still being requested at each time. The guidelines from Redis on the subject https://redis.io/topics/sentinel-clients advises to get the master addr only on first initialization, reconnection, failure cases.

**Comment by pmlopes** [comment]
@adoignies thanks for the detailed report. I am thinking that we need to add tests to verify if we are using a pool and a connection is in pub/sub mode, when the connection is recycled that it leaves that mode (or is forcefully closed).

**Comment by pmlopes** [comment]
@adoignies I'm having trouble to reproduce this. I've created the test:

```java
  @Test
  public void testGetClientToMasterAndDontLeakPushNotifications(TestContext should) {
    final Async test = should.async();
    final AtomicInteger cnt = new AtomicInteger(10);

    Redis redis = Redis.createClient(
        rule.vertx(),
        new RedisOptions()
          .setType(RedisClientType.SENTINEL)
          .addConnectionString("redis://localhost:5000")
          .addConnectionString("redis://localhost:5001")
          .addConnectionString("redis://localhost:5002")
          .setMasterName("sentinel7000")
          .setRole(RedisRole.MASTER));

    rule.vertx()
      .setPeriodic(1000L, t -> {
        redis.send(Request.cmd(Command.INFO))
          .onSuccess(res -> {
            if (cnt.getAndDecrement() == 0) {
              test.complete();
            }
          })
          .onFailure(should::fail);
      });
  }
```

But after the 10 times it didn't fail. Can you make a reproducer?

**Comment by adoignies** [comment]
The behaviour of the SentinelClient seems to create a connection with subcribe command and then put it to the connections pool. When all the pool has been filled it grabbed an already used connection on which a subscribe command is being played, and play a get-master-addr-by-name which resulted in an error on redis side. Then the same process is repeated for the next sentinel nodes.
So it fails after N times, N depends on the max pool size and the number of configured nodes.
In my case, i configured 2 nodes and it failed after 10 success. 
Since you configure 3 nodes, it should failed after 15 success maybe a bit more.

**Comment by pmlopes** [comment]
@adoignies I'm still having trouble to reproduce the issue; however, I think it should be fixed with the PR mentioned above. Could you give it a try and see if it works for you?

**Comment by adoignies** [comment]
@pmlopes Hello, I retrieved the branch. Unfortunately it failed after the 6st request.

the 6th request never get the response, the async handler"createAndConnect" in createConnectionInternal method l.120 is never called back and the get-master-addr-by-name command not called this time. I will try to figure out why the handler is not called back.

It sounds weird that you are unable to reproduce it. To compare the sentinel configuration can you please give details on how you instantiate your sentinels instances ? On my side i used Master/Slave and 3 sentinels instances installed on severals VM.

**Comment by pmlopes** [comment]
Can you try to make a simple app running against the container mentioned in the read me?

**Comment by adoignies** [comment]
Yes I will try asap. I used redis 5.0 version. Maybe the redis version can explain the behaviour. I will try to build a container with 5.0 if i can't reproduce the issue as you.

**Comment by adoignies** [comment]
Hello Paulo, 
I reproduce the same behavior from a starter with a simple MainVerticle and the container.
It fails after 8 requests.
Maybe a warn log on the get-master-addr-by-name callback in failure case will help you to detect the error.

`public class MainVerticle extends AbstractVerticle {

    @Override
    public void start(Promise<Void> startPromise) throws Exception {
        RedisOptions redisOptions = new RedisOptions().setType(RedisClientType.SENTINEL);

        Redis redis = Redis.createClient(
                vertx,
                redisOptions
                        .setMasterName("sentinel7000")
                        .addConnectionString("redis://localhost:5000")
                        .addConnectionString("redis://localhost:5001")
                        .addConnectionString("redis://localhost:5002")
                        .setRole(RedisRole.MASTER)
                        .setProtocolNegotiation(true));


        vertx.createHttpServer().requestHandler(req -> {
            req.pause();
            redis.send(Request.cmd(Command.INFO)).onComplete(info -> {
                req.resume();
                req.response()
                        .putHeader("content-type", "text/plain")
                        .end("Hello from Vert.x!" + info.failed());
            });

        }).listen(8888, http -> {
            if (http.succeeded()) {
                startPromise.complete();
                System.out.println("HTTP server started on port 8888");
            } else {
                startPromise.fail(http.cause());
            }
        });
    }
}`


I have a slightly different behavior because of the container https://github.com/Grokzen/docker-redis-cluster. 
But the same error is send back by redis. Indeed only the sentinel on port 5000 monitors the sentinel7000 master.
The client fails in a NullPointerException when trying to get the master from the 2nd sentinel on port 5001.

**Comment by adoignies** [comment]
Sorry Paulo, i try to compile the merged fix, but i get a compilation error on RedisConnectionManager :

Compilation failure
[ERROR] /mnt/c/PROJECT/OML/vertx-redis-client/src/main/java/io/vertx/redis/client/impl/RedisConnectionManager.java:[397,20] incompatible types: io.vertx.core.impl.ContextInternal cannot be converted to io.vertx.core.impl.EventLoopContext

Can you please remind me the maven repository where i can find the 4.2.0 snapshot version of vertx dependencies ?

Thanks in advance!

**Comment by pmlopes** [comment]
@adoignies you can find precompiled snapshots on the repo: `https://s01.oss.sonatype.org/content/repositories/snapshots/` just update your maven/gradle to use those snapshots to test

**Comment by adoignies** [comment]
@pmlopes Thanks for your answer. I did the test, the behavior is fixed but i've got a side effect.

The number of connections to the sentinel keeps increasing at each request, and only one connection is opened to the master.

Here some networks investigations, which show the numbers of connections opened :
``
sudo netstat -antp | grep -v LISTEN | grep -c 7000
1
sudo netstat -antp | grep -v LISTEN | grep -c 5000
53
``

Regarding the connections used to listen to switch-master event, i think that only one connection for this matter is enough for the sentinel client.  The PUBSUB command seems to be played at each request, maybe it's not the initial intention of the sentinel client ?

**Comment by pmlopes** [comment]
@adoignies This is a design issue. Initially users were expected to get a connection from the client. Since this connection would live forever the sentinel client runs the 2nd connection in parallel. In 4.0 we have a connection pool so users don't worry about managing the connection. And this has the side effect that in this mode the sentinel connection is constantly restarted.

We need to decouple the sentinel to the client in order to have a single watcher.

A similar issue is needed in clustered mode too as we don't need to get the cluster slots on each connection.

I think we should create a new issue to implement these improvements.
